include/libopencm3/stm32/fdcan.h

```diff
@@ -0,0 +1,793 @@
+
+LGPL License Terms @ref lgpl_license
+*/
+/*
+ * This file is part of the libopencm3 project.
+ *
+ * Copyright (C) 2021 Eduard Drusa <ventyl86@netkosice.sk>
+ *
+ * This library is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Lesser General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This library is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this library.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+
+
+#ifndef LIBOPENCM3_FDCAN_H
+#define LIBOPENCM3_FDCAN_H
+
+#include <libopencm3/stm32/memorymap.h>
+#include <libopencm3/cm3/common.h>
+
+/** @{ */
+
+/* FDCAN block base addresses. Used in functions to identify FDCAN block being manipulated. */
+
+/** @defgroup fdcan_block FDCAN block base addresses
+ * @{
+ */
+#define CAN1							FDCAN1_BASE
+#define CAN2							FDCAN2_BASE
+#define CAN3							FDCAN3_BASE
+/**@}*/
+
+
+/** @defgroup fdcan_fifo Named constants for FIFOs
+ * @{
+ */
+#define FDCAN_FIFO0						0
+#define FDCAN_FIFO1						1
+/**@}*/
+
+
+/** @defgroup FDCAN registers file in each FDCAN block. */
+
+#define FDCAN_CREL(can_base)			MMIO32(can_base + 0x0000)
+#define FDCAN_ENDN(can_base)			MMIO32(can_base + 0x0004)
+#define FDCAN_DBTP(can_base)			MMIO32(can_base + 0x000C)
+#define FDCAN_TEST(can_base)			MMIO32(can_base + 0x0010)
+#define FDCAN_RWD(can_base)				MMIO32(can_base + 0x0014)
+#define FDCAN_CCCR(can_base)			MMIO32(can_base + 0x0018)
+#define FDCAN_NBTP(can_base)			MMIO32(can_base + 0x001C)
+#define FDCAN_TSCC(can_base)			MMIO32(can_base + 0x0020)
+#define FDCAN_TSCV(can_base)			MMIO32(can_base + 0x0024)
+#define FDCAN_TOCC(can_base)			MMIO32(can_base + 0x0028)
+#define FDCAN_TOCV(can_base)			MMIO32(can_base + 0x002C)
+#define FDCAN_ECR(can_base)				MMIO32(can_base + 0x0040)
+#define FDCAN_PSR(can_base)				MMIO32(can_base + 0x0044)
+#define FDCAN_TDCR(can_base)			MMIO32(can_base + 0x0048)
+#define FDCAN_IR(can_base)				MMIO32(can_base + 0x0050)
+#define FDCAN_IE(can_base)				MMIO32(can_base + 0x0054)
+#define FDCAN_ILS(can_base)				MMIO32(can_base + 0x0058)
+#define FDCAN_ILE(can_base)				MMIO32(can_base + 0x005C)
+#define FDCAN_RXGFC(can_base)			MMIO32(can_base + 0x0080)
+#define FDCAN_XIDAM(can_base)			MMIO32(can_base + 0x0084)
+#define FDCAN_HPMS(can_base)			MMIO32(can_base + 0x0088)
+#define FDCAN_RXF0S(can_base)			MMIO32(can_base + 0x0090)
+#define FDCAN_RXF0A(can_base)			MMIO32(can_base + 0x0094)
+#define FDCAN_RXF1S(can_base)			MMIO32(can_base + 0x0098)
+#define FDCAN_RXF1A(can_base)			MMIO32(can_base + 0x009C)
```

Comment:
this for instance could potentially beocme 
```
FDCAN_RXFIFOS(can, fifo)  MMIO(can_base + 0x0090 + (8*fifo))
FDCAN_RXFIFOA(can, fifo)  MMIO(can_base + 0x0094 + (8*fifo))
```
then in a few places later, where you have two if cases with ~identical code, you could simplify it a bit.  Just a thought though

Reply:
That's actually a good idea! I felt, that the code below, which gets affected by such change was ugly and duplicated but couldn't come up with elegant way to deduplicate it.